.github/actions/nexus-promote-staging/action.yml

```diff
@@ -0,0 +1,83 @@
+name: Promote Nexus staging repository
+description: Close and promote a Nexus Staging Repository
+inputs:
+  version:
+    description: Version being released. Used to update the repository description while closing and releasing
+    required: true
+  staging-repository:
+    description: The id of the staging repository to be promoted
+    required: true
+  nexus-username:
+    description: The Nexus username
+    required: true
+  nexus-password:
+    description: The Nexus password
+    required: true
+  nexus-url:
+    description: The base URL to the Nexus server
+    required: false
+    default: "https://artifacts.alfresco.com/nexus"
+runs:
+  using: composite
+  steps:
+    - name: Close staging repository
+      shell: bash
+      env:
+        NEXUS_USERNAME: ${{ inputs.nexus-username }}
+        NEXUS_PASSWORD: ${{ inputs.nexus-password }}
+        NEXUS_URL: ${{ inputs.nexus-url }}
+        VERSION: ${{ inputs.version }}
+        STAGING_REPOSITORY_ID: ${{ inputs.staging-repository }}
+      run: |
+        CLOSE_STAGING_PAYLOAD_CONTENT=$(envsubst < $GITHUB_ACTION_PATH/close-payload-template.json)
+        curl -u "${NEXUS_USERNAME}":"${NEXUS_PASSWORD}" \
+          -H "Accept: application/json" \
+          -H "Content-Type: application/json" \
+          -d "$CLOSE_STAGING_PAYLOAD_CONTENT" \
+          "${NEXUS_URL}"/service/local/staging/bulk/close
+
+        wait_until_true () {
+            local attempt_counter=0
+            local max_attempts=12
+
+            until "$@"
+            do
+              if [ ${attempt_counter} -eq ${max_attempts} ]
+              then
+                echo "Max attempts reached"
+                exit 1
+              fi
+
+              attempt_counter=$((attempt_counter+1))
+              echo "Condition not reached yet. Attempt $attempt_counter out of $max_attempts. Retrying..."
+              sleep 5
+            done
+        }
+
+        staging_repo_is_closed () {
+          curl -u "${NEXUS_USERNAME}":"${NEXUS_PASSWORD}" \
+            -X GET "${NEXUS_URL}"/service/local/staging/profile_repositories \
+            | yq -p=xml e '.stagingRepositories.data.stagingProfileRepository[]| select (.repositoryId == env(STAGING_REPOSITORY_ID)) | .type ' \
+            | grep open > /dev/null 2>&1
+        }
+
+        # the close operation is asynchronous and before proceeding with the release step, we need to make sure the repository is closed
+        echo "Waiting for staging repository $STAGING_REPOSITORY_ID to get closed..."
+        wait_until_true staging_repo_is_closed
+        echo "Repository $STAGING_REPOSITORY_ID successfully closed!"
+
+    - name: Release staging repository
```

Comment:
The action name does not make it explicit that a release is also performed. We do not have any use case to promote without releasing, but maybe it would make sense to move release to a dedicated action, wdyt?
Also, it would allow skipping the async check if not needed (as the release operaton is async too: maybe the completion should be checked on it too?)

Reply:
In all the use cases we have now, we always need to close and release, but I can split it into too actions and add option to wait for the close one:
```yaml
     - uses: Alfresco/alfresco-build-tools/.github/actions/nexus-close-staging@ref
        with:
          wait-for-completion: true
          version: ${{ inputs.version }}
          staging-repository: ${{ inputs.staging-repository}}
          nexus-username: ${{ secrets.NEXUS_USERNAME }}
          nexus-password: ${{ secrets.NEXUS_PASSWORD }}
    
     - uses: Alfresco/alfresco-build-tools/.github/actions/nexus-release-staging@ref
        with:
          version: ${{ inputs.version }}
          staging-repository: ${{ inputs.staging-repository}}
          nexus-username: ${{ secrets.NEXUS_USERNAME }}
          nexus-password: ${{ secrets.NEXUS_PASSWORD }}
```

Concerning the wait mechanism, I only added for the first one because we need to make sure the staging repository is closed before calling the release action. BTW, the naming is not consistent between the UI and the related REST API, as in the UI the `Release` button calls the `promote` API.
For the second one, the wait mechanism is not strictly needed, but it could be added for consistency and if we want to make sure that the repository was actually been released. The check however, would mean verifying that the repository has disappeared from the list of repositories.

What do you think?

Reply:
Ok, I'll update it. Thanks for the feedback.